File: ts/earthbucks-lib/src/block-builder.ts
import { Block } from "./block.js";
import { Header } from "./header.js";
import { Tx } from "./tx.js";
import type { Script } from "./script.js";
import { TxIn } from "./tx-in.js";
import { TxOut } from "./tx-out.js";
import type { FixedBuf } from "@webbuf/fixedbuf";
import { U8 } from "@webbuf/numbers";
import type { U16BE, U128BE, U256BE } from "@webbuf/numbers";
import { U32BE } from "@webbuf/numbers";
import type { U64BE } from "@webbuf/numbers";
import { MerkleTree } from "./merkle-tree.js";

export class BlockBuilder {
  public header: Header;
  public txs: Tx[];
  public rootMerkleTree: MerkleTree;

  constructor(header: Header, txs: Tx[], rootMerkleTree: MerkleTree) {
    this.header = header;
    this.txs = txs;
    this.rootMerkleTree = rootMerkleTree;
  }

  static fromBlock(block: Block): BlockBuilder {
    const header = block.header;
    const txs = block.txs;
    const merkleRoot = MerkleTree.fromLeafHashes(txs.map((tx) => tx.id()));
    return new BlockBuilder(header, txs, merkleRoot);
  }

  static fromGenesis(
    initialTarget: U256BE,
    outputScript: Script,
    outputAmount: U64BE,
  ): BlockBuilder {
    const txInput = TxIn.fromMintTxScript(outputScript);
    const txOutput = new TxOut(outputAmount, outputScript);
    const mintTx = new Tx(new U8(0), [txInput], [txOutput], new U32BE(0n));
    const txs = [mintTx];
    const merkleRoot = MerkleTree.fromLeafHashes(txs.map((tx) => tx.id()));
    const merkleRootId: FixedBuf<32> = merkleRoot.hash as FixedBuf<32>;
    const header = Header.fromGenesis(merkleRootId, initialTarget);
    return new BlockBuilder(header, txs, merkleRoot);
  }

  toBlock(): Block {
    return new Block(this.header, this.txs);
  }

  addTx(tx: Tx): BlockBuilder {
    const txs = [...this.txs, tx];
    const merkleRoot = this.rootMerkleTree.addLeafHash(tx.id());
    const merkleRootId = merkleRoot.hash as FixedBuf<32>;
    const header = this.header.addTx(merkleRootId);
    return new BlockBuilder(header, txs, merkleRoot);
  }
}

File: rs/earthbucks_lib/src/block_builder.rs
use crate::block::Block;
use crate::header::Header;
use crate::merkle_txs::MerkleTxs;
use crate::script::Script;
use crate::tx::Tx;
use crate::tx_in::TxIn;
use crate::tx_out::TxOut;

pub struct BlockBuilder {
    pub header: Header,
    pub txs: Vec<Tx>,
    pub merkle_txs: MerkleTxs,
}

impl BlockBuilder {
    pub fn new(header: Header, txs: Vec<Tx>, merkle_txs: MerkleTxs) -> Self {
        Self {
            header,
            txs,
            merkle_txs,
        }
    }

    pub fn from_block(block: Block) -> Self {
        let header = block.header;
        let txs = block.txs;
        let merkle_txs = MerkleTxs::new(txs.clone());
        Self::new(header, txs, merkle_txs)
    }

    pub fn from_genesis(output_script: Script, output_amount: u64, new_timestamp: u64) -> Self {
        let mut header = Header::from_genesis(new_timestamp);
        let tx_input = TxIn::from_coinbase(output_script.clone());
        let tx_output = TxOut::new(output_amount, output_script.clone());
        let coinbase_tx = Tx::new(0, vec![tx_input], vec![tx_output], 0);
        let txs = vec![coinbase_tx];
        let merkle_txs = MerkleTxs::new(txs.clone());
        let root: [u8; 32] = merkle_txs.root;
        header.merkle_root = root;
        Self::new(header, txs, merkle_txs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::block::Block;
    use crate::header::Header;
    use crate::numbers::u256;
    use crate::tx::Tx;

    #[test]
    fn test_from_block() {
        let bh = Header {
            version: 0,
            prev_block_id: [0; 32],
            merkle_root: [0; 32],
            timestamp: 0,
            block_num: 0,
            target: u256::from(0u8),
            nonce: u256::from(0u8),
            work_ser_algo: 0,
            work_ser_hash: [0; 32],
            work_par_algo: 0,
            work_par_hash: [0; 32],
        };
        let tx = Tx::new(0, vec![], vec![], 0);
        let block = Block::new(bh.clone(), vec![tx]);
        let bb = BlockBuilder::from_block(block);
        assert_eq!(bb.header.version, bh.version);
        assert_eq!(bb.header.prev_block_id, bh.prev_block_id);
        assert_eq!(bb.header.merkle_root, bh.merkle_root);
        assert_eq!(bb.header.timestamp, bh.timestamp);
        assert_eq!(bb.header.target, bh.target);
    }
}
